synapse/rest/media/v1/_base.py

```diff
@@ -284,13 +285,14 @@ async def respond_with_responder(
     finish_request(request)
 
 
-class Responder:
+class Responder(ABC):
     """Represents a response that can be streamed to the requester.
 
     Responder is a context manager which *must* be used, so that any resources
     held can be cleaned up.
     """
 
+    @abstractmethod
```

Comment:
This *must* be implemented by storage providers, correct? (Otherwise nothing makes sense.)

I think this won't break the S3 storage provider, which is the only one I know of:

https://github.com/matrix-org/synapse-s3-storage-provider/blob/c65f9e5ba538f0aa0e16229656121a3cc783d82e/s3_storage_provider.py#L303-L311

Reply:
Yeah. If you don't override this method then we won't ever write to the IConsumer here https://github.com/matrix-org/synapse/blob/f3519679b167ff342d00ba3b9756169cd3e43399/synapse/rest/media/v1/_base.py#L274 or here https://github.com/matrix-org/synapse/blob/2dff93099b5aa7e213da76a9c4b3de84385b58e1/synapse/rest/media/v1/media_storage.py#L257. I think that would mean you present a piece of media that is a 0-length byte sequence. 
I think that's an error.

(If someone disagrees, they can override the method to return `None` as today.)